openbb_sdk/providers/alpha_vantage/openbb_alpha_vantage/models/stock_eod.py

```diff
@@ -0,0 +1,217 @@
+        and that the required fields are not None if the function is required.
+        """
+
+        fields = cls.__fields__
+        function = values["function"]
+
+        for field in fields:
+            if (
+                available_on_functions := fields[field].field_info.extra.get(
+                    "available_on_functions", None
+                )
+            ) and function not in available_on_functions:
+                values[field] = None
+            if (
+                (
+                    required_on_functions := fields[field].field_info.extra.get(
+                        "required_on_functions", None
+                    )
+                )
+                and function in required_on_functions
+                and values[field] is None
+            ):
+                raise ValueError(f"Field {field} is required on function {function}")
+
+        return values
+
+    @validator("month")
+    def month_validate(cls, v):  # pylint: disable=E0213
+        """Validate month, check if the month is in YYYY-MM format."""
+        if v is not None:
+            try:
+                datetime.strptime(v, "%Y-%m")
+            except ValueError as e:
+                raise e
+        return v
+
+
+class AlphaVantageStockEODData(StockEODData):
+    """Alpha Vantage Stock End of Day Data."""
+
+    class Config:
+        """Pydantic alias config using fields dict."""
+
+        fields = {"date": "timestamp", "adj_close": "adjusted_close"}
+
+    adjusted_close: PositiveFloat = Field(
+        description=DATA_DESCRIPTIONS.get("adj_close", "")
+    )
+    dividend_amount: NonNegativeFloat = Field(
+        description="Dividend amount paid for the corresponding date.",
+    )
+    split_coefficient: NonNegativeFloat = Field(
+        description="Split coefficient for the corresponding date.",
+    )
+
+
+class AlphaVantageStockEODFetcher(
+    Fetcher[
+        AlphaVantageStockEODQueryParams,
+        List[AlphaVantageStockEODData],
+    ]
+):
+    """Transform the query, extract and transform the data from the Alpha Vantage endpoints."""
+
+    @staticmethod
+    def transform_query(params: Dict[str, Any]) -> AlphaVantageStockEODQueryParams:
+        """Transform the query."""
+
+        transformed_params = params
+
+        now = datetime.now().date()
+        if params.get("start_date") is None:
+            transformed_params["start_date"] = now - relativedelta(years=1)
+
+        if params.get("end_date") is None:
+            transformed_params["end_date"] = now
+
+        return AlphaVantageStockEODQueryParams(**transformed_params)
+
+    @staticmethod
+    def extract_data(
+        query: AlphaVantageStockEODQueryParams,
+        credentials: Optional[Dict[str, str]],
+        **kwargs: Any,
+    ) -> dict:
+        """Return the raw data from the Alpha Vantage endpoint."""
+
+        api_key = credentials.get("alpha_vantage_api_key") if credentials else ""
+
```

Comment:
Here you can convert the `period` to upper case, attach it to `TIME_SERIES` string.

You can also check if the functions and the parameters are correct and raise the Exception. This is optional because for the wrong parameters, there will be an exception raised by the AV API.

Reply:
Would go for rely only on the validators and pydantic's own data validation and let the htpp response we'll get from AV do the rest.

Reply:
Good idea!